Source/Fuse.Storage/UserSettingsImpl.uno

```diff
@@ -0,0 +1,269 @@
+using Uno;
+using Uno.IO;
+using Uno.UX;
+using Uno.Collections;
+using Uno.Compiler.ExportTargetInterop;
+
+namespace Fuse.Storage
+{
+	interface IUserSettings
+	{
+		string GetStringValue(string key);
+
+		double GetNumberValue(string key);
+
+		bool GetBooleanValue(string key);
+
+		void SetStringValue(string key, string value);
+
+		void SetNumberValue(string key, double value);
+
+		void SetBooleanValue(string key, bool value);
+
+		void Remove(string key);
+
+		void Clear();
+	}
```

Comment:
So, I'm not a huge fan of using interfaces to hide platform differences. The reason is that interfaces have a relatively large performance penalty (lots of vtable work), and we don't have a proper devirtualizer in Uno.

So I think it's better to do stuff like

```C#
if defined(iOS)
    iOSImpl.DoStuff();
else if defined(Android)
    AndroidImpl.DoStuff();
else
    build_error;
```

Yes, it's uglier, and yes, it's just papering over a less than ideal compiler, but I doubt that last part will be fixed anytime soon, and the performance problem just keeps adding up everywhere if we don't work actively to avoid it.

Reply:
It should be fairly easy to change this code to `abstract class` instead of `interface`, since there's no need for multiple inheritance here.

On `abstract class` all function pointers are known at compile-time and performance is good (comparable to calling a function via pointer in native C). Calling methods on `interface` will involve first looking-up the vtable, which adds a little more run-time overhead. In this case, it's essentially one type-check and return a pointer to the first vtable, since only one interface is used. `interface` performance in Uno isn't a disaster I think, but `abstract class` is faster, and plain non-virtual methods (as kusma suggests) are fastest (by a only small margin I think, and more ugly...).

Reply:
I think the overhead depends on how often you'd call the code. For something like this, I'd expect quite a few calls in common cases, so I think we should consider keeping the overhead as low as possible.

Abstract class has pretty bad cache-behavior. These kinds of things are also often the reason why OOP programs tend to be more sluggish than other languages, as things like this adds up all over the software-stack. If the mind-set is to do the known-performant thing by default, you're going to get much less overhead-buildup.

Reply:
This is all true, but when looking at the bigger picture we see that the code already involves a bunch of overhead we have little control over:

1. The underlying implementations are invoking platform-specific Java or Obj-C APIs, where I'm fairly confident both languages already involve more overhead regarding method calls than is being introduced by the Uno-layer.

2. And ultimately, clients will interact with this code via the JavaScript-engine, adding even more run-time overhead.

I'd argue that even if we make the Uno code as fast as possible, the win will be relatively very small compared to the overhead in 1 and 2, and we can't avoid bad cache-behavior.

So, I just wanted to suggest an alternative solution that may possibly fix 80% of the problem by 20% of the engineering-effort, simply switching from `interface` to `abstract class`. That said, the solution you're suggesting is definitely the fastest one.

Reply:
Right. Yeah, perhaps the bridging into the other languages will dominate the performance...

Reply:
Wow, there's so many information and insight with this discussion. Thank you, I'll try to fix the code using the fastest approach mention by @kusma